Framework/Geometry/test/StructuredDetectorTest.h

```diff
@@ -0,0 +1,210 @@
+    det->setPos(1000., 2000., 3000.);
+
+    std::vector<double> x{0, 1, 2, 0, 1, 2, 0, 1, 2};
+    std::vector<double> y{0, 0, 0, 1, 1, 1, 2, 2, 2};
+
+    // Initialize with these parameters
+    det->initialize(2, 2, x, y, 0, true, 2, 1);
+
+    do_test_on(det);
+
+    // --- Now make a parametrized version ----
+    ParameterMap_sptr pmap(new ParameterMap());
+    StructuredDetector *parDet = new StructuredDetector(det, pmap.get());
+
+    do_test_on(parDet);
+
+    delete det;
+    delete parDet;
+  }
+
+  void testIncorrectVertexArraySize() {
+    boost::shared_ptr<Geometry::Object> cuboidShape =
+        ComponentCreationHelper::createCuboid(0.5);
+
+    StructuredDetector *det = new StructuredDetector("MyStructuredDetector");
+    det->setPos(1000., 2000., 3000.);
+
+    std::vector<double> x{0, 1, 2, 0, 1, 2};
+    std::vector<double> y{0, 0, 0, 1, 1, 1};
+
+    // Initialize with these parameters
+    TS_ASSERT_THROWS(det->initialize(2, 2, x, y, 0, true, 2, 1),
+                     std::invalid_argument);
+
+    x.resize(3);
+
+    TS_ASSERT_THROWS(det->initialize(2, 2, x, y, 0, true, 2, 1),
+                     std::invalid_argument);
+
+    x.resize(0);
+    y.resize(0);
+
+    TS_ASSERT_THROWS(det->initialize(2, 2, x, y, 0, true, 2, 1),
+                     std::invalid_argument);
+
+    delete det;
+  }
+
+  /** Test on a structured detector that will be
+  * repeated on an un-moved parametrized version.
+  */
+  void do_test_on(StructuredDetector *det) {
+    TS_ASSERT_EQUALS(det->xpixels(), 2);
+    TS_ASSERT_EQUALS(det->ypixels(), 2);
+
+    auto size = (det->xpixels() + 1) * (det->ypixels() + 1);
+
+    TS_ASSERT_EQUALS(det->getXValues().size(), size);
+    TS_ASSERT_EQUALS(det->getYValues().size(), size);
+
+    // Go out of bounds
+    TS_ASSERT_THROWS(det->getAtXY(-1, 0), std::runtime_error);
+    TS_ASSERT_THROWS(det->getAtXY(0, -1), std::runtime_error);
+    TS_ASSERT_THROWS(det->getAtXY(5, 0), std::runtime_error);
+    TS_ASSERT_THROWS(det->getAtXY(0, 6), std::runtime_error);
+
+    // Check some ids
+    TS_ASSERT_EQUALS(det->getAtXY(0, 0)->getID(), 0);
+    TS_ASSERT_EQUALS(det->getAtXY(0, 1)->getID(), 1);
+    TS_ASSERT_EQUALS(det->getAtXY(1, 1)->getID(), 3);
+
+    std::pair<int, int> xy;
+    int x;
+    int y;
+
+    TS_ASSERT_THROWS_NOTHING(xy = det->getXYForDetectorID(0));
+    x = xy.first;
+    y = xy.second;
+    TS_ASSERT_EQUALS(x, 0);
+    TS_ASSERT_EQUALS(y, 0);
+
+    TS_ASSERT_THROWS_NOTHING(xy = det->getXYForDetectorID(1));
+    x = xy.first;
+    y = xy.second;
+    TS_ASSERT_EQUALS(x, 0);
+    TS_ASSERT_EQUALS(y, 1);
+
+    TS_ASSERT_THROWS_NOTHING(xy = det->getXYForDetectorID(2));
+    x = xy.first;
+    y = xy.second;
+    TS_ASSERT_EQUALS(x, 1);
+    TS_ASSERT_EQUALS(y, 0);
+
+    // Name
+    TS_ASSERT_EQUALS(det->getAtXY(0, 1)->getName(),
+                     "MyStructuredDetector(0,1)");
+    TS_ASSERT_EQUALS(det->getChild(1)->getName(), "MyStructuredDetector(x=1)");
```

Comment:
Good testing! Did you know that cxx test also provides message versions of most of their macros. For example `TS_ASSERT_EQUALS` has a cousin `TSM_ASSERT_EQUALS`, which allows you to provide a message that will be displayed in the case of failure. I find that quite useful, especially when it allow me to avoid the debugger.